src/DynamoCore/Models/WorkspaceModel.cs

```diff
@@ -786,6 +794,51 @@ internal void IncrementPasteOffset()
             this.currentPasteOffset = (this.currentPasteOffset + PASTE_OFFSET_STEP) % PASTE_OFFSET_MAX;
         }
 
+        internal void SetWorkspaceToState(PresetState state)
+        {
+            if (state == null)
+            {
+                Log("Attempted to apply a PresetState that was null");
+                return;
+            }
+            //start an undoBeginGroup
+            using (var undoGroup = this.undoRecorder.BeginActionGroup())
+            {
+               
+               //reload each node, and record each each modification in the undogroup
+                foreach (var node in state.Nodes)
+                {
+                    //check that node still exists in this workspace, 
+                    //otherwise bail on this node
+                    if (nodes.Contains(node))
+                    {
+                        var originalpos = node.Position;
+                        var serializedNode = state.SerializedNodes.ToList().Find(x => Guid.Parse(x.GetAttribute("guid")) == node.GUID);
+                        //overwrite the xy coords of the serialized node with the current position, so the node is not moved
+                        serializedNode.SetAttribute("x", originalpos.X.ToString());
+                        serializedNode.SetAttribute("y", originalpos.Y.ToString());
+
+                        this.undoRecorder.RecordModificationForUndo(node);
+                        this.ReloadModel(serializedNode);
+                        
+                    }
+                }
+                //select all the modified nodes in the UI
+                DynamoSelection.Instance.ClearSelection();
+                state.Nodes.ToList().ForEach(x => DynamoSelection.Instance.Selection.Add(x));
```

Comment:
Just use a `foreach` here over the `IEnumerable<NodeModel>`.  Otherwise, you're artificially introducing a new list.